### ISSUE TITLE
DAFSA needs Runtime Conversions

### ISSUE THREAD

**Issue by darnir** [rockdaboot/libpsl#42]
I just saw the commit cd5a11 which states that DAFSA requires runtime conversions. In light of this, can we have DAFSA as a compile time option and retain the old library code as well? For systems where you want only builtin data with no runtime conversions.

We already have all the code, it's only a matter of separating it out into #ifdef blocks

**Comment by rockdaboot** [comment]
If we can modify the dafsa structure so that it could contain utf-8, we could leave away runtime idna and still would have pretty small data footprint.

**Comment by darnir** [comment]
That would work too. But will require some effort to change the DAFSA structure. I'll take a look when I can.

**Comment by rockdaboot** [comment]
We could convert domains with 8bit chars (utf-8) into base64, label by label. Much less work than changing the basics. And maybe even less data footprint. Wdyt ?

**Comment by darnir** [comment]
We could do that. Though I'm not entirely convinced that it'll be the best idea for the long term.

As I see it currently, we'd have to maintain a patchset that needs to be applied to any updates in the PSL. Though, it's entirely possible that I misunderstood you.

**Comment by rockdaboot** [comment]
Functions like psl_is_public_suffix() accept either a lowercase UTF-8 domain name or it's punycode representation (e.g. müller.de and xn--mller-kva.de). The built-in data so far had both representations included. But now with DAFSA, we are limited to ASCII characters. To test müller.de we would need to convert it into punycode within the psl_ functions. But without runtime IDNA conversion, we can't do that.

I see these options:

1. Enrich the builtin data with a second (easy to do) ASCII representation (e.g. base64). Calling psl_is_public_suffix(müller.de) first does the base64 encoding and then the lookup within the DAFSA builtin data. Caveat: there may be conflicts with existing domain labels.

2. Embed a to-punycode converter, which basically means copying that routine from one of the libidn/libicu libraries into libpsl. I would like to avoid this code duplication, also libpsl would be blown up to some extend.

3. Extend DAFSA to work with 8bit data. No idea how much work that is.

4. Don't support --disable-runtime.

**Comment by rockdaboot** [comment]
I decided for 2. by taking the punycode generation code from the original example code foudn at http://www.nicemice.net/idn/punycode-spec.gz. It is what libidn and libidn2 use (did not look at libicu). 

The copyright is a very lax one, so it fits IMO.

But I found libpsl being linked with libicu when using --disable-runtime which is not needed. It needs some tweaking in configure.ac + Makefile.am.

**Comment by liljenzin** [comment]
I just want to add my comments in case the issue is reopened in the future.

The dafsa format uses one bit in each byte to mark end of string. When I wrote the original code I saw this as an opportunity to avoid storing the trailing null byte, since all data was in 7-bit ASCII anyway. But since suffixes are compressed it is not a big saving after all. Allowing 8-bit characters would be a minor change to the format.

I would avoid base64 encoding since it would ruin the suffix compression. For example, all names in the .com domain are now joined so that the common suffix is stored only once, but with base64 encoding they would all get different suffixes that can't be joined.

**Comment by rockdaboot** [comment]
@liljenzin You are the original author ? Then I have to say that in many years of code review, your code was the only one that made me say 'wow' (in a positive meaning) ! :smiley: 

The base64 encoding was meant to do label by label, so that DAFSA encoding would still join the suffixes. But that is history.

If you could change the code to deal with 8 bit strings, I could remove the punycode converter from the library. As @daurminator mentioned somewhere, that code does not normalize the input strings (which AFAIK would pull in much more code and/or data tables). So we have a possible problem here, that could be solved by an amended (8 bit) DAFSA encoding. The caveat is that we likely can't use the registered 'file magic' (```b'.DAFSA@PSL_0   \n'``` - V0), we have to register a new version (V1) of that. The new lookup code will be incompatible with V0 DAFSA files, so libpsl has to provide two lookup routines (maybe not so bad as one might think).

**Comment by liljenzin** [comment]
@rockdaboot I changed the python script to work with 8-bit input and the difference in output size is extremely small. So that part is an easy fix.

Do we have to support different versions of data? I thought the look up code was linked with a fixed data set, but I guess you added support to inject external data if this is an issue? But it might be possible to detect the actual format and support both versions in the lookup code. (I thought of adding a null byte first, as the old format can never start with that.)

**Comment by rockdaboot** [comment]
@liljenzin 
> work with 8-bit input and the difference in output size is extremely small

Thanks, that sounds very promising. Fast work :+1: 

Just to make clear (for occasional readers): currently we have to use the punycode representation of international domain rules. The lookup function needs to convert international domain names (IDN) into punycode (IDNA).

The new 8-bit format could store both, IDN and punycode representations. So the lookup function would not need to convert IDN into IDNA any more. The main advantage is less code complexity for lookups. Especially when libpsl is built without IDNA libraries (libidn/libdn2/libicu), we don't to need carry the discrete IDNA code any more. Less code = shorter library, better maintainability and faster lookups for IDNs.

> Do we have to support different versions of data?

The creation (psl-make-dafsa) does not need to support both versions (just the new 8-bit format).
The lookup code has to support both versions - and your idea with a first 0 byte to distinguish between the two formats seems to be perfect !
Indeed, libpsl is assumed to load external DAFSA blobs used for lookups. Supporting both formats is a *must*. IMO, that could automatically be done in `src/lookup_string_in_fixed_set.c`.

If you are going to create a pull request, feel free to add your name as author/creator to the appropriate files, also to the AUTHORS file.

**Comment by liljenzin** [comment]
I will get time to make a patch next week. Can we assume UTF-8 or will other encoding be used as well?

**Comment by rockdaboot** [comment]
Utf-8 only

**Comment by liljenzin** [comment]
I have created a pull request providing UTF-8 support.

Given the input with national characters:
```
www.å.se, 0
www.ä.se, 0
```

The following dafsa will be generated:
```
L00: 0x81  # -> L01
L01: 0x77  # match "w"
L02: 0x77  # match "w"
L03: 0x77  # match "w"
L04: 0x2E  # match "."
L05: 0x1F  # match multibyte prefix
L06: 0xC3  # match "C" (Shifted UTF byte 1/2)
L07: 0x02  # -> L09
L08: 0x82  # -> L10
L09: 0xE5  # match "e" (Shifted UTF byte 2/2)
L10: 0x82  # -> L12
L11: 0x64  # match "d" (Shifted UTF byte 2/2)
L12: 0x2E  # match "."
L13: 0x73  # match "s"
L14: 0x65  # match "e"
L15: 0x80  # return 0

```
The special byte value 0x1F was not used in the original format and therefore multibyte characters will always mismatch if an old parser is used, otherwise the new format is fully compatible with an old parser and pure ASCII strings should match as before.

A new parser supports the old format automatically as the new format is an extension to the old format.

**Comment by rockdaboot** [comment]
That sounds awesome, very well thought of.
I am currently in the middle of a dozen small things to finish... but will have a look ASAP.

**Comment by liljenzin** [comment]
Take your time. This time no one else has reviewed the changes, so please ask if you have questions or if you find something could be done in a better way.

**Comment by rockdaboot** [comment]
I created a new branch 'utf-8-dafsa' with your both commits merged into one.

:-) And the first question is: How to distinguish between old and new DAFSA format ?
I need that in src/psl.c/psl_load_fp() and store that information into 'psl_ctx_t' for use in _psl_is_public_suffix(). Currently, I convert UTF-8 into punycode and use that for lookup - but knowing the 'psl' points to the new DAFSA, I could skip that conversion. If 'psl' is an old DAFSA, I have to convert.

**Comment by liljenzin** [comment]
That is a good question and the short answer is that it is not possible by inspecting the dafsa only, since input without utf-8 characters will generate exactly the same dafsa as before. I thought you were going to remove the punycode converter so that conversion was no longer an option.

What options do we have? We could add a version number after the last data byte in the dafsa. Data always ends with a byte in the range [0x80-0x8F] and we could add a byte outside this range to indicate the new version. That is a small change.

**Comment by rockdaboot** [comment]
> I thought you were going to remove the punycode converter so that conversion was no longer an option.

Yes in the long term. In the short term we have to stay compatible in both directions, something you already have considered very well. E.g. the new code loads an old DAFSA file (without utf-8, just punycode) - how should we lookup IDN without punycode conversion ? 
But if we load a new DAFSA, I would skip the punycode conversion and lookup using the domain I got from the user. And at some time in the future we can drop support for the old DAFSA version and throw out the punycode code completely.

> Data always ends with a byte in the range [0x80-0x8F] and we could add a byte outside this range to indicate the new version.

Perfect ! I simply check the last byte read by psl_load_fp() and voila. Could you make up a commit with that change and push it to your branch... I should get a notification than and start extending the tests.

**Comment by liljenzin** [comment]
Since we will support two formats simultaneously it must also be possible to test both formats, thus I have added encoding switches to psl-make-dafsa. The original format is default and UTF-8 support has to be enabled by "--encoding=utf-8" on the command line. Maybe we should switch default mode if ASCII mode will be deprecated?

I pushed a separate patch that makes the punycode conversion depending on the loaded dafsa version. Otherwise I couldn't test that it works. But you might want to drop that patch and do it some other way.

**Comment by rockdaboot** [comment]
> Maybe we should switch default mode if ASCII mode will be deprecated?

IMO, we should promote the new format, so default should be utf-8 mode.

When I just looked at your patches, I saw that we have 'space' in the file header:
```
return b'.DAFSA@PSL_0   \n' + words_to_whatever(words, bytearray)
```
We could also use the byte before `\n` as a flag, setting it to 0x01 - I did quickly did that with a hex editor to test if the `file` tool still recognizes the file. It does.
```
$ file tests/psl.dafsa 
tests/psl.dafsa: Public Suffix List data (optimized) (Version 0)
```

Wouldn't the code / doc changes not be simpler if we do that instead of appending 0x01 ?
Just a thought, WDYT ?

**Comment by rockdaboot** [comment]
FYI, the current state is:
I made things up in my branch utf-8-dafsa, added a few things, set utf-8 mode as default for psl-make-dafsa. Also 'make' now creates an ASCII DAFSA and a UTF-8 DAFSA in tests/.
I fixed psl-make-dafsa/parse-psl to add the UTF-8 rules to the DAFSA.
Please review.

First tests look good though, I'll make up the test suite using both DAFSA files in the next days.
I also need a check with the old code + new DAFSA file (just sayin' to not forget it myself).

**Comment by liljenzin** [comment]
> We could also use the byte before \n as a flag, setting it to 0x01 - I quickly did that with a hex editor to test if the file tool still recognizes the format. It does.

Wow! I didn't know you had patched the file command. :+1:
> Wouldn't the code / doc changes not be simpler if we do that instead of appending 0x01 ?

That is absolutely fine with me. I never thought of this possibility since I wanted to put the same information in the C/C++ output that still is supported by the script. But it is probably less useful since the used format should be known when building the program in that case.

>  I made things up in my branch utf-8-dafsa, added a few things, set utf-8 mode as default for psl-make-dafsa. Also 'make' now creates an ASCII DAFSA and a UTF-8 DAFSA in tests/.
I fixed psl-make-dafsa/parse-psl to add the UTF-8 rules to the DAFSA.

Sounds good. I will have a look.

**Comment by rockdaboot** [comment]
> First tests look good though, I'll make up the test suite using both DAFSA files in the next days.
I also need a check with the old code + new DAFSA file (just sayin' to not forget it myself).

Done

> Wow! I didn't know you had patched the file command. :+1:

Daniel Kahn Gillmor did it.

> ...I wanted to put the same information in the C/C++ output that still is supported by the script.
But it is probably less useful since the used format should be known when building the program in that case.

Hmmm, have to check that. Maybe a small change needed.

Basically we are done, I guess. What a cool enhancement, especially the compatibility of the new format with old lookup code ! You made an awesome piece of code !

**Comment by liljenzin** [comment]
>  Hmmm, have to check that. Maybe a small change needed.

Do you want me to change the code? I can revert the old version patch and make a new one, but then it will not work before we add new code that reads the version from the header instead. And then it is not obvious to me where to best do that in psl.c.
> Basically we are done, I guess. What a cool enhancement, especially the compatibility of the new format with old lookup code ! You made an awesome piece of code !

It was quite fun to do this patch actually. This is the first time someone asked for a change since the original version three years ago and I almost forgot about it.

**Comment by rockdaboot** [comment]
> Do you want me to change the code?

No, let's leave it as it is. It is fine and functional.

> It was quite fun to do this patch actually. This is the first time someone asked for a change since the original version three years ago and I almost forgot about it.

That reminds me of thinking how we can convince Chromium and Firefox devs to use libpsl... Are you working for Google/Chrome ? Maybe you have an idea how to get it in there. My next steps are a new libpsl release and getting it uploaded to Debian (wget depends on it).

**Comment by liljenzin** [comment]
> That reminds me of thinking how we can convince Chromium and Firefox devs to use libpsl... Are you working for Google/Chrome ? Maybe you have an idea how to get it in there.

Ryan Sleevi might be a better person to ask regarding chromium since he actually works for Google. But they might be hard to convince since they already have their code that os doing exactly what they need. Personally I am working at Opera with their Android browser, which is also based on chromium, but strategic decisions in chromium are always taken by Google.
